tensorflow/lite/micro/micro_context.h

```diff
@@ -0,0 +1,138 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_LITE_MICRO_MICRO_CONTEXT_H_
+#define TENSORFLOW_LITE_MICRO_MICRO_CONTEXT_H_
+
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/micro/micro_allocator.h"
+#include "tensorflow/lite/micro/micro_graph.h"
+
+namespace tflite {
+// MicroContext is eventually going to become the API between TFLM and the
+// kernels, replacing all the functions in TfLiteContext. The end state is code
+// kernels to have code like:
+//
+// MicroContext* micro_context = GetMicroContext(context);
+// micro_context-><TFLM kernel API>
+class MicroContext {
+ public:
+  // Does not take any ownership, and all pointers must refer to valid objects
+  // that outlive the one constructed.
+  explicit MicroContext(MicroAllocator* allocator, const Model* model,
+                        MicroGraph* graph);
+
+  // Allocate persistent buffer which has the same life time as the interpreter.
+  // Returns nullptr on failure.
+  // The memory is allocated from from tail.
+  // This method is only available in Init or Prepare stage.
+  // Virtual so that it can be mocked for kernel tests.
+  // WARNING: This is an experimental interface that is subject to change.
+  virtual void* AllocatePersistentBuffer(size_t bytes);
+
+  // Request a scratch buffer in the arena through static memory planning.
+  // This method is only available in Prepare stage and the buffer is allocated
+  // by the interpreter between Prepare and Eval stage. In Eval stage,
+  // GetScratchBuffer API can be used to fetch the address.
+  // Virtual so that it can be mocked for kernel tests.
+  // WARNING: This is an experimental interface that is subject to change.
+  virtual TfLiteStatus RequestScratchBufferInArena(size_t bytes,
+                                                   int* buffer_idx);
+
+  // Get the scratch buffer pointer.
+  // This method is only available in Eval stage.
+  // Virtual so that it can be mocked for kernel tests.
+  // WARNING: This is an experimental interface that is subject to change.
+  virtual void* GetScratchBuffer(int buffer_idx);
+
+  // Returns a TfLiteTensor struct for a given index.
+  // Virtual so that it can be mocked for kernel tests.
+  // WARNING: This is an experimental interface that is subject to change.
+  virtual TfLiteTensor* GetTensor(int tensor_idx);
+
+  // Returns a TfLiteEvalTensor struct for a given index.
+  // Virtual so that it can be mocked for kernel tests.
+  // WARNING: This is an experimental interface that is subject to change.
+  virtual TfLiteEvalTensor* GetEvalTensor(int tensor_idx);
+
+  // Accesses external contexts by type.
```

Comment:
not sure what this means.

Also, let's make a note that return value can be a nullptr if ExternalContext is not set.

I think we can go without documenting the accessor functions and inlining them so that the code documents the function.